### PR TITLE
fix(medusa): ensure transaction is reused when service calls itself

### PR DESCRIPTION
**Pull request by srindom**
**What**
- ProductService.create calls ProductService.retrieve before returning. This fix ensures that the manager created in the atomicPhase is used when ProductService.retrieve is called.

**Why**
- Without the explicit use of the same manager in the retrieve call, race conditions easily occur if you have concurrent ProductService.create calls. The code below can reproduce the scenario:
```javascript
  const productData = [
    {
      barcode: `1234233423-${Math.random() * 100}`,
      external_id: `1234233423-${Math.random() * 100}`,
      description: "super cool product",
      discountable: true,
      hs_code: "1234213",
      origin_country: "DK",
      title: `Super cool product ${Math.random() * 100}`,
      type: { value: "Eyewear" },
      sales_channels: [{ id: sc.id }],
    },
    {
      barcode: `1234233423-${Math.random() * 100}`,
      external_id: `random-external-id-${Math.random() * 100}`,
      description: "super cool product",
      discountable: true,
      hs_code: "1234213",
      origin_country: "DK",
      title: `Super cool product ${Math.random() * 100}`,
      type: { value: "Eyewear" },
      sales_channels: [{ id: sc.id }],
    },
  ];

  const result = await Promise.all(
    productData.map(async (product) => productService.create(product))
  );
```

**Explaination**

What happens is the following:

- Request 1 calls `ProductService.create`. This in turn calls `atomicPhase` which sets the `ProductService.transactionManager_ = txForReqOne`
- Request 1 creates the product in the DB with `txForReqOne`.
- Request 2 calls `ProductService.create`. This in turn calls `atomicPhase` which sets the `ProductService.transactionManager_ = txForReqTwo`
- Request 1 reaches the end of `ProductService.create` where `this.retrieve` is called. Because the ProductService is a singleton the retrieve call will attempt to use `txForReqTwo` to fetch the product. 
- **Error**: since `txForReqTwo` can't read the data of `txForReqOne`, the product is not found.

### CODE REVIEW

**Comment by adrien2p** [comment]
Nice catch 💪 you found that probably because there is somewhere where the service is not called within a transaction right?

**Comment by olivermrbl** [comment]
Should we fix this across the entire service? Think we use this pattern in a couple of other places

**Comment by adrien2p** [comment]
> Should we fix this across the entire service? Think we use this pattern in a couple of other places

Is it also the only service or should we check them all?

**Review by riqwan — APPROVED** [review]

**Comment by olivermrbl** [comment]
@adrien2p, I will migrate the rest of the ProductService before we merge, but will keep the rest (if any) for a separate PR

**Comment by adrien2p** [comment]
Sounds good to me, if the pr is not done just after we might need to create a ticket?

**Comment by srindom** [comment]
@olivermrbl, @adrien2p  - can we apply automerge or is there still more here?

**Comment by adrien2p** [comment]
I ll do a final review, haven't seen the last commit. Butnif you want we can auto merge and if i have any comments we can handle them separately to not block it

File: packages/medusa/src/services/product.ts [hunk 1]
```diff
@@ -594,7 +595,7 @@ class ProductService extends TransactionBaseService {
           )
       }
 
-      const result = await this.retrieve(product.id, {
+      const result = await this.withTransaction(manager).retrieve(product.id, {
```
Comment (adrien2p) [inline packages/medusa/src/services/product.ts:598]: todo: line 559 is missing. maybe we can do something like `productServiceTx = this.withTransaction(manager)` and then use it in both places?

File: packages/medusa/src/services/product.ts [hunk 2]
```diff
@@ -833,7 +834,7 @@ class ProductService extends TransactionBaseService {
         )
       }
 
-      const result = await this.retrieve(productId)
+      const result = await this.withTransaction(manager).retrieve(productId)
```
Comment (adrien2p) [inline packages/medusa/src/services/product.ts:837]: suggestion: same as the previous one to not recreate the instance multiple times everytime

**Review by adrien2p — APPROVED** [review]
LGTM 🔥 couple of comments

File: packages/medusa/src/services/product.ts (hunk 1 above)
Comment (adrien2p) [reply packages/medusa/src/services/product.ts:598]: @srindom I ve seen that you have merged but there is still a missing transaction here

**Pull request merged**